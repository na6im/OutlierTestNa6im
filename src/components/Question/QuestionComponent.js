import React, { useState, useEffect } from 'react'
import {
  Category,
  Question,
  QuestionCore,
  AnswerButton,
  Result
} from './styles'
import fullStar from '../../blackStar.png'
import star from '../../greyStar.png'

function QuestionComponent ({
  question,
  nextQuestionHandler,
  onAnswerHandler,
  end
}) {
  const [answers, setAnswers] = useState([])
  const [answered, setAnswered] = useState()

  const difficulty = {
    easy: 1,
    medium: 2,
    hard: 3
  }

  useEffect(() => {
    let answerOptions = question.incorrect_answers
    answerOptions.splice(
      Math.floor(Math.random() * Math.floor(3)),
      0,
      question.correct_answer
    )
    setAnswers(answerOptions)
    setAnswered()
  }, [question])

  function getAnswerColor (answer) {
    if (answer === question.correct_answer) return 'black'
    if (answer === answered) return 'white'
  }

  function renderResult () {
    return answered === question.correct_answer ? 'Correct' : 'Sorry!'
  }

  function onClickHandler (answer) {
    if (!answered) {
      setAnswered(answer)
      onAnswerHandler(answer === question.correct_answer)
    }
  }

  return (
    <React.Fragment>
      <Category>{decodeURIComponent(question.category)}</Category>
      <div>
        {[...Array(3)].map((x, id) => (
          <img
            alt={'star'}
            key={id}
            src={id < difficulty[question.difficulty] ? fullStar : star}
          />
        ))}
      </div>
      <Question>
        <QuestionCore style={{ gridArea: 'question' }}>
          {decodeURIComponent(question.question)}
        </QuestionCore>
        {answers.map((answer, id) => (
          <AnswerButton
            key={id}
            id={id}
            disabled={
              !!answered &&
              answer !== question.correct_answer &&
              answer !== answered
            }
            onClick={() => {
              onClickHandler(answer)
            }}
            answerColor={!!answered && getAnswerColor(answer)}
          >
            {decodeURIComponent(answer)}
          </AnswerButton>
        ))}
      </Question>
      {!!answered && (
        <Result>
          {!end ? renderResult() : 'its the end'}
          <AnswerButton
            style={{
              width: 'fit-content',
              padding: '5px 1em',
              marginTop: '2em'
            }}
            onClick={() => nextQuestionHandler()}
          >
            {!end ? 'Next question' : 'Play again'}
          </AnswerButton>
        </Result>
      )}
    </React.Fragment>
  )
}

export default QuestionComponent
