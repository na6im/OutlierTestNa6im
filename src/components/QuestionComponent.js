import React, { useState, useEffect } from 'react'
import { Title, Category, Question, QuestionCore, AnswerButton } from './styles'
import star from '../blackStar.png'

function QuestionComponent ({ question }) {
  const [answers, setAnswers] = useState([])
  const [answered, setAnswered] = useState()

  useEffect(() => {
    const answerOptions = question.incorrect_answers
    answerOptions.splice(
      Math.floor(Math.random() * Math.floor(3)),
      0,
      question.correct_answer
    )
    setAnswers(answerOptions)
  }, [])
  {
    //TODO stars
  }

  function getAnswerColor (answer) {
    if (answer === question.correct_answer) return 'black'
    if (answer === answered) return 'white'
  }

  return (
    <React.Fragment>
      <Title>Question 16 of 20</Title>
      <Category>Entertaiment: Board Games</Category>
      <div>
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
      </div>
      <Question>
        <QuestionCore style={{ gridArea: 'question' }}>
          {decodeURIComponent(question.question)}
        </QuestionCore>
        {answers.map((answer, id) => (
          <AnswerButton
            id={id}
            disabled={
              !!answered &&
              answer !== question.correct_answer &&
              answer !== answered
            }
            onClick={() => setAnswered(answer)}
            answerColor={!!answered && getAnswerColor(answer)}
          >
            {decodeURIComponent(answer)}
          </AnswerButton>
        ))}
      </Question>
    </React.Fragment>
  )
}

export default QuestionComponent
