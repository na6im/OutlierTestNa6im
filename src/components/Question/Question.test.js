import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import QuestionComponent from './QuestionComponent'
import '@testing-library/jest-dom/extend-expect'

test('renders corectly', () => {
  const nextQuestionHandler = jest.fn()
  const onAnswerHandler = jest.fn()
  const question = {
    category: 'Entertainment%3A%20Video%20Games',
    type: 'multiple',
    difficulty: 'hard',
    question:
      'What%20was%20the%20name%20of%20the%20hero%20in%20the%2080s%20animated%20video%20game%20%27Dragon%27s%20Lair%27%3F',
    correct_answer: 'Dirk%20the%20Daring',
    incorrect_answers: ['Arthur', 'Sir%20Toby%20Belch', 'Guy%20of%20Gisbourne']
  }
  const { getAllByTestId, getAllByAltText, getByText } = render(
    <QuestionComponent
      question={question}
      nextQuestionHandler={nextQuestionHandler}
      onAnswerHandler={onAnswerHandler}
      end={false}
    />
  )

  expect(getAllByTestId('answer').length).toBe(4)
  expect(getAllByAltText('star').length).toBe(3)
  expect(getByText(decodeURIComponent(question.question))).toBeInTheDocument()
})

describe('<Question />', () => {
  beforeEach(() => {
    nextQuestionHandler.mockReset()
    onAnswerHandler.mockReset()
    onReset.mockReset()
  })

  const nextQuestionHandler = jest.fn()
  const onAnswerHandler = jest.fn()
  const onReset = jest.fn()
  const question = {
    category: 'Entertainment%3A%20Video%20Games',
    type: 'multiple',
    difficulty: 'hard',
    question:
      'What%20was%20the%20name%20of%20the%20hero%20in%20the%2080s%20animated%20video%20game%20%27Dragon%27s%20Lair%27%3F',
    correct_answer: 'Dirk%20the%20Daring',
    incorrect_answers: ['Arthur', 'Sir%20Toby%20Belch', 'Guy%20of%20Gisbourne']
  }

  it('Expect to not log errors in console', () => {
    const { getByText } = render(
      <QuestionComponent
        question={question}
        nextQuestionHandler={nextQuestionHandler}
        onAnswerHandler={onAnswerHandler}
        end={false}
      />
    )

    let correct = getByText(decodeURIComponent(question.correct_answer))
    fireEvent.click(correct)
    expect(onAnswerHandler).toHaveBeenCalledTimes(1)
    expect(getByText('Correct')).toBeInTheDocument()
    expect(getByText('Next question')).toBeInTheDocument()
  })

  it('should render Sorry! on incorrect Answer', () => {
    const { getByText } = render(
      <QuestionComponent
        question={question}
        nextQuestionHandler={nextQuestionHandler}
        onAnswerHandler={onAnswerHandler}
        end={false}
      />
    )

    let incorrect_answers = getByText(
      decodeURIComponent(question.incorrect_answers[0])
    )
    fireEvent.click(incorrect_answers)
    expect(onAnswerHandler).toHaveBeenCalledTimes(1)
    expect(getByText('Sorry!')).toBeInTheDocument()
    expect(getByText('Next question')).toBeInTheDocument()
  })

  it('Answering Question and asking for next', () => {
    const { getByText } = render(
      <QuestionComponent
        question={question}
        nextQuestionHandler={nextQuestionHandler}
        onAnswerHandler={onAnswerHandler}
        end={false}
      />
    )

    let incorrect_answers = getByText(
      decodeURIComponent(question.incorrect_answers[0])
    )
    fireEvent.click(incorrect_answers)

    const nextQuestion = getByText('Next question')
    expect(getByText('Sorry!')).toBeInTheDocument()
    expect(nextQuestion).toBeInTheDocument()
    expect(onAnswerHandler).toHaveBeenCalledTimes(1)

    fireEvent.click(nextQuestion)
    expect(nextQuestionHandler).toHaveBeenCalledTimes(1)
  })

  it('Showing Play again at the end', () => {
    const { getByText } = render(
      <QuestionComponent
        question={question}
        nextQuestionHandler={nextQuestionHandler}
        onAnswerHandler={onAnswerHandler}
        onReset={onReset}
        end={true}
      />
    )

    let incorrect_answers = getByText(
      decodeURIComponent(question.incorrect_answers[0])
    )
    fireEvent.click(incorrect_answers)

    expect(getByText('its the end')).toBeInTheDocument()
    expect(onAnswerHandler).toHaveBeenCalledTimes(1)
    fireEvent.click(getByText('Play again'))
    expect(onReset).toHaveBeenCalledTimes(1)
  })

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = render(
      <QuestionComponent
        question={question}
        nextQuestionHandler={nextQuestionHandler}
        onAnswerHandler={onAnswerHandler}
        end={false}
      />
    )
    expect(firstChild).toMatchSnapshot()
  })
})
