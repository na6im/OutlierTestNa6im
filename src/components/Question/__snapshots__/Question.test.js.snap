// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> Should render and match the snapshot 1`] = `
<div
  class="sc-AxjAm dwNmLq"
>
  Entertainment: Video Games
</div>
`;
